libdrgn/python/program.c

```diff
@@ -477,6 +477,73 @@ static struct drgn_error *py_object_find_fn(const char *name, size_t name_len,
 	return drgn_object_copy(ret, &((DrgnObject *)obj)->obj);
 }
 
+static struct drgn_error *py_symbol_find_fn(const char *name, uint64_t addr,
+					    enum drgn_find_symbol_flags flags,
+					    void *data, struct drgn_symbol_result_builder *builder)
+{
+	PyGILState_guard();
+
+	_cleanup_pydecref_ PyObject *name_obj = NULL;
+	if (flags & DRGN_FIND_SYM_NAME) {
+		name_obj = PyUnicode_FromString(name);
+		if (!name_obj)
+			return drgn_error_from_python();
+	} else {
+		name_obj = Py_None;
+		Py_INCREF(name_obj);
+	}
+
+	_cleanup_pydecref_ PyObject *address_obj = NULL;
+	if (flags & DRGN_FIND_SYM_ADDR) {
+		address_obj = PyLong_FromUnsignedLong(addr);
+		if (!address_obj)
+			return drgn_error_from_python();
+	} else {
+		address_obj = Py_None;
+		Py_INCREF(address_obj);
+	}
+
+	_cleanup_pydecref_ PyObject *one_obj = PyBool_FromLong(flags & DRGN_FIND_SYM_ONE);
+
+	_cleanup_pydecref_ PyObject *tmp = PyObject_CallFunction(data, "OOO", name_obj,
+								 address_obj, one_obj);
+	if (!tmp)
+		return drgn_error_from_python();
+
+	_cleanup_pydecref_ PyObject *obj =
+		PySequence_Fast(tmp, "Symbol find callback must return a sequence");
+	if (!obj)
+		return drgn_error_from_python();
+
+	size_t len = PySequence_Fast_GET_SIZE(obj);
+	if (len > 1 && (flags & DRGN_FIND_SYM_ONE))  {
+		return drgn_error_create(DRGN_ERROR_INVALID_ARGUMENT,
+					 "symbol find callback returned multiple elements, but one was requested");
```

Comment:
Should this use `PyErr_SetString(PyExc_ValueError, ...)` and `drgn_error_from_python()`?

Reply:
In fact, I tried this and then [decided it was not great](https://github.com/osandov/drgn/pull/241#discussion_r1367556859). That basically invents a Python exception which never happened, converts it into a Drgn error, passes it down into the symbol finder layer, and then typically it gets translated back into a Python exception again (e.g. if you called `prog.symbol()`). But the translated exception is not still a ValueError, it's a generic `Exception` with the `ValueError: ...` text embedded within it.

Reply:
Ah, thanks for the reminder. Yeah, there's some messiness with `drgn_error_from_python()` that I still need to figure out, so this is fine.